### PR TITLE
NV-1302 feat(ws): upgrade socket.io v4

### PR DESCRIPTION
**Pull request by vinhnguyen1211**
### What change does this PR introduce? 

[NV-1302] 🚀 Feature: Upgrade to socket.io-client v4
https://github.com/novuhq/novu/issues/2175

### Why was this change needed?

In case project also use `socket.io-client` but has newer version then need to install 2 `socket.io-client` package

### Other information (Screenshots)

- I upgrade `@nestjs/platform-socket.io` and use `ioredis` to create Redis instance because it support `keyPrefix` while `redis` doesn't  (https://github.com/redis/node-redis/issues/1657)
- Also upgrade `socket.io-client` in notification-center

### CODE REVIEW

File: apps/ws/src/shared/framework/redis.adapter.ts [hunk 1]
```diff
@@ -1,27 +1,23 @@
...
-import * as redisIoAdapter from 'socket.io-redis';
+import { ServerOptions } from 'socket.io';
+import Redis from 'ioredis';
+import { createAdapter } from '@socket.io/redis-adapter';
 import { getRedisPrefix } from '@novu/shared';
 
 export class RedisIoAdapter extends IoAdapter {
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  createIOServer(port: number, options?: any): any {
+  createIOServer(port: number, options?: ServerOptions): any {
```
Comment (scopsy) [inline apps/ws/src/shared/framework/redis.adapter.ts:9]: Love the add types here!

File: libs/embed/package.json [hunk 2]
```diff
@@ -94,6 +94,7 @@
     "@commitlint/cli": "^7.1.2",
     "@commitlint/config-conventional": "^7.1.2",
     "@rollup/plugin-replace": "^2.4.2",
+    "@types/estree": "^1.0.0",
```
Comment (scopsy) [inline libs/embed/package.json:97]: Was this one missing? not sure for what package it was related to

Comment (vinhnguyen1211) [reply libs/embed/package.json]: you're right, I just remove it

**Review by p-fernandez — CHANGES_REQUESTED** [review]
Thanks a lot for this, it is very helpful.
I will only request you to also update `apps/widget` as it is missing to upgrade the Socket.io client from version `2.3.1`.

File: apps/ws/package.json [hunk 3]
```diff
@@ -28,33 +28,35 @@
...
-    "@nestjs/platform-socket.io": "^7.6.17",
+    "@nestjs/platform-socket.io": "^9.2.1",
     "@nestjs/swagger": "^4.8.0",
     "@nestjs/terminus": "^7.2.0",
     "@nestjs/websockets": "^7.6.17",
     "@novu/dal": "^0.9.1",
     "@novu/shared": "^0.9.0",
     "@novu/testing": "^0.9.1",
     "@sentry/node": "^6.3.3",
+    "@socket.io/redis-adapter": "^7.2.0",
```
[the remark this answers is not in the document]
Comment (vinhnguyen1211) [reply apps/ws/package.json:39]: yes, old package is deprecated https://www.npmjs.com/package/socket.io-redis

**Review by p-fernandez — APPROVED** [review]
🌟

**Comment by vinhnguyen1211** [comment]
@p-fernandez thanks for the approval.
I saw 1 check is failed, do I need to do something about it? Also how can I get the PR merge?

**Comment by p-fernandez** [comment]
> @p-fernandez thanks for the approval. I saw 1 check is failed, do I need to do something about it? Also how can I get the PR merge?

It is a new code quality integrity check we are testing. I will bypass it as we need to think if we want to tune it in a way. Don't worry about it. Thanks again.

**Pull request merged**